src/main/java/com/microsoft/sqlserver/jdbc/SQLServerPreparedStatement.java

```diff
@@ -170,7 +171,13 @@ String getClassNameInternal() {
         cacheKey = new CityHash128Key(sql);
 
         // Parse or fetch SQL metadata from cache.
-        ParsedSQLCacheItem parsedSql = getOrCreateCachedParsedSQL(cacheKey, sql);
+        ParsedSQLCacheItem parsedSql = getCachedParsedSQL(cacheKey, sql);
+        if (null != parsedSql) {
+            isExecutedAtLeastOnce = true;
+        }
+        else {
+            parsedSql = parseAndCacheSQL(cacheKey, sql);
+        }
```

Comment:
Ok, after wasting my time with my own analysis I concluded you are right @brettwooldridge :-) 

What I completely missed is that the cache hit ratio with ``enablePrepareOnFirstPreparedStatementCall=true`` is 1/queriesPerCacheSize given the unique query usage is uniformly random (i.e. 100% where cache size is at least as large as total number of queries and 50% where the cache size is half of the number of unique queries). On the other hand the default of ``enablePrepareOnFirstPreparedStatementCall=false`` is 1/queriesPerCacheSize^2 (i.e. where cache size is half the number of queries the hit rate would be 25%).

When using your updated change it is close to the same as ``enablePrepareOnFirstPreparedStatementCall=false`` while still using sp_executesql when unique queries are executed. Consider me completely on-board with the chance of having isExecutedAtLeastOnce be dependent on the parsed SQL text cache :-) 

I have pushed the change to metadata caching and hope to get it into the release :-)

Thanks for being persistent @brettwooldridge!

PS I believe I fixed the final bug for the not-yet-GitHub:ed tests :-)